### ISSUE TITLE
[Bug]: FC0004 wrong sorting

### ISSUE THREAD

**Issue by guidorobben** [ALCops/Analyzers#245]
### Analyzer

FormattingCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

FC0004

### Description

The sorting of the permissionset differs from the AZ Dev Tools sorting. Is AZ doing it wrong or ALCops?

The code below is showing warning FC0004. If I change the order of the lines its gone. 

### AL code to reproduce

```al
permissionset 50100 GeneratedPermission
{
    Assignable = true;
    Permissions =
        codeunit "Post. Appr. Setup Wiz Hlp" = X,
        codeunit "Post Inv. Appr. Setup Hlp" = X;
}
```

### Additional context

How this rule is currently, we probably will turn it off because the sorting of AZ can be done during the cleanup function. This is one action to clean everything. So this will be preferred to manually changing the permissionset. 

0.8.0 beta3

**Comment by guidorobben** [comment]
This also happens for permissions in objects.

**Comment by Arthurvdv** [comment]
I have a PR in place to try to mimic the specifics of the AZ AL Dev Tools by @anzwdev, but when I had give it some more thought I'm doubting on merging this.

When I look at an example [EmailLoggingObj.PermissionSet.al](https://github.com/microsoft/ALAppExtensions/blob/main/Apps/W1/EmailLogging/app/src/permissions/EmailLoggingObj.PermissionSet.al) this sorting seems like just alphabetical. Should this then be te standard? The diagnostics of the FormattingCop also provides a Code Action (QuickFix) to fix the sorting.

For the specific sorting I couldn't find anything on [alguidelines.dev](https://alguidelines.dev) of this. Would be great to know what the best practice would be (that's why I also tagged Andrzej here) what the motive behind the specific sorting of the AZ AL Dev Tools is.

**Comment by guidorobben** [comment]
I think it's wise to wait until its clear what to do. That's why I asked: _Is AZ doing it wrong or ALCops?_

For usit doesn't matter if we go left or right, as long as both act the same. I have no preference for how it is sorted.

**Comment by Aize-S** [comment]
As stated in #372 I agree that we would want to sort the objects alphabetically. But I don't want to sort the object types alphabetically. I want to sort them "the old fashioned" way:
- Table
- TableData
- Report
- Codeunit
- XMLPort
- Page
- Query

In regards to the alphabetical sorting of the objects I agree with @guidorobben's last comment; it doesn't matter which way we go.

**Comment by Arthurvdv** [comment]
## FC0004 permission sorting — investigation of the current state, and a call for consensus

Following up on this issue. There is **no official, documented specification** for how the
`Permissions` property should be sorted, and the tools in the ecosystem disagree. I think the
right way forward is for us to **agree as a community on what the sorting should be**, rather
than ALCops unilaterally picking one. This comment lays out the current state so we can
discuss from facts. **It is not a proposal for a specific order** — it's a starting point.

### 1. The tools disagree with each other

I verified the sort logic from source of the two tools people actually use to produce
permission sets:

**A. AL Language / NAV SDK — "Generate permission set as AL object"**
(`PermissionSetGenerator.GetMissingPermissionObjects`) sorts by
`ObjectType` (the `PermissionObjectType` enum), then object name (`OrdinalIgnoreCase`), then
object id. The enum values are `TableData=0, Table=1, Report=3, Codeunit=5, XmlPort=6,
Page=8, Query=9, System=10`, so the **type order is**:

> tabledata, table, report, codeunit, xmlport, page, query, system

with `tabledata` and `table` as separate blocks and names compared **ordinally**
(`"Item 10" < "Item 2"`).

**B. AZ AL Dev Tools — "Sort Permissions"**
(`PermissionComparer` in `anzwdev/al-code-outline`) uses the priority map
`table=0, tabledata=1, codeunit=2, page=3, query=4, report=5, xmlport=6`. `table` and
`tabledata` are **one combined group, interleaved and sorted by name** (table before
tabledata only on an exact name tie), always before the other types; the rest are ordered by
the map. Names use a **natural/alphanumeric** comparison (`"Item 2" < "Item 10"`, spaces
stripped, `InvariantCultureIgnoreCase`). So the **type order is**:

> (table+tabledata by name), codeunit, page, query, report, xmlport

**C. Current FC0004** sorts type keywords **alphabetically** (`codeunit, page, query, report,
table, tabledata, xmlport`) with ordinal names — which matches **neither** A nor B.

These three are mutually incompatible on three independent axes:

| Axis | A. AL Language / SDK | B. AZ Dev Tools | C. Current FC0004 |
|---|---|---|---|
| table vs tabledata | tabledata block, then table block | interleaved by name | tabledata after table |
| Non-table order | report, codeunit, xmlport, page, query | codeunit, page, query, report, xmlport | codeunit, page, query, report, xmlport |
| Name comparison | ordinal | natural/alphanumeric | ordinal |
| `system` | last | table group | by keyword |

(For context: PR #251 implemented option B and was closed without merging.)

### 2. What does Microsoft actually do? (BC 28, W1)

I ran a script over the BC 28 W1 source (`MSDyn365BC.Sandbox.Code.History`,
branch `w1-28`), looking only at `permissionset` / `permissionsetextension` objects and
ignoring the `Permissions` property on other object types.

**Corpus:** 17,370 AL files → 1,150 permissionset objects (295 of them
`permissionsetextension`); 727 have a `Permissions` property; **607** have ≥2 entries (the
ones that can reveal an ordering).

**Findings:**

- **All 607 reference objects by name** — none by numeric id.
- **0 use namespace-qualified names** anywhere in the `Permissions` property.
- **83% (502/607) contain only a single object type**, so they don't constrain the *type*
  ordering at all. Only 105 have ≥2 types, and only **22 structurally distinguish A from B**.
- Among those **22 discriminating sets**: **5 match A (SDK) only, 8 match B (AZ) only, 8
  match neither, 1 matches both.** In other words, **Microsoft is not internally
  consistent** — different apps use different orders.
- **Within-type name order:** ~55% are alphabetically sorted (333 satisfy both ordinal and
  natural; +13 ordinal-only), but **~41% (250/607) are not name-sorted at all**. These are
  the generated `*Objects`/`*Obj` permission sets (e.g.
  `AITestToolkitObj`, `ShpfyObjects`, `ExcelReportsObjects`, `BCPerfToolkitObj`), whose
  entries appear to follow **object-id / declaration order** rendered as names — which any
  by-name rule (A, B, or C) would flag as "unsorted".
- **Natural vs ordinal names matters in only a handful of cases** (11 sets sort naturally but
  not ordinally), so it's a real but rare differentiator.
- Sets containing **both** `table` and `tabledata` are rare (5); 2 interleaved (B-style), 3
  separated (A-style) — too few to conclude.

**Illustrative generated set** (`AI Test Toolkit - Obj`): type order is table → codeunit →
xmlport → page (enum/A order), but the table names are clearly **not alphabetical** — they're
in object-id order.

### 3. Namespaces

For completeness, since it came up: the AL Language / SDK secondary sort key is the **bare
object name** and ignores the namespace; two objects sharing a short name across namespaces
are then separated only by object id. In the entire BC 28 W1 codebase, **zero** permission
entries are namespace-qualified, so this is a non-issue for first-party code today. It could
still matter for PTE/AppSource code that qualifies names — if we standardize on a rule, we
should decide whether the namespace participates in the sort (I'd lean toward "no, bare name
only", matching the SDK).

### 4. Where I'd like the community's input

- **Which order should FC0004 enforce** — A (AL Language / SDK), B (AZ Dev Tools), the
  "old fashioned" table-first order from the comment above, or something else? Given that even
  Microsoft is inconsistent, there's no obvious "correct" answer, so I'd like agreement before
  changing the rule.
- **Should FC0004 tolerate the generated `*Objects` sets** (object-id order), or should those
  be reformatted by-name too?
- **Are there other sources we should check** before deciding — other localizations/versions,
  other formatters/linters, `alguidelines.dev`, or additional first-party repos? Pointers
  welcome.

---
_Generated with Claude Opus 4.8_

**Comment by MODUSCarstenScholling** [comment]
I'd go with option B (AZ Dev Tools) because it's easier to read and manage.

It groups the `table` and `tabledata` permissions together. This makes it much simpler to look through permission sets. You can quickly check if both permissions are there and what actions are allowed without having to look in different places.

The order also seems more natural, following how people usually think about these objects:

1.  `table` and `tabledata`, grouped and sorted by name
2.  `codeunit`
3.  `page`
4.  `query`
5.  `report`
6.  `xmlport`
7.  `system`

When names are the same, I'd put `table` before `tabledata`. Sorting them naturally and without regard to case just makes more sense, like putting `Item 2` before `Item 10`.

Option A might match the SDK, but its order is mostly based on internal enum values. While it makes sense technically, it's not very helpful when you're reading or checking permission sets. Option C is straightforward, but sorting by object type alphabetically doesn't have much practical meaning and separates related permissions.

Microsoft's own documentation/implementation doesn't offer a clear standard. It seems like just copying Microsoft isn't always the best approach, especially when they use different methods themselves. For FC0004, focusing on readability, predictability, and making reviews efficient should be more important than following an inconsistent pattern.

The question remains whether to change FC0004's default behavior or add a new setting for how things are sorted. I'd prefer the order that makes more sense to people to be the default, but having a configuration option could help projects that are already set up to follow the SDK or another convention.

**Comment by guidorobben** [comment]
I vote for B.

**Comment by Aize-S** [comment]
My C/AL wired dinosaur heart wants the object type sorting of A. 

But my brain also votes for option B, so let's go with that.

**Comment by frottke** [comment]
Option B, for sure.

**Comment by Arthurvdv** [comment]
I think we have a winner! 😅

**Comment by Arthurvdv** [comment]
I’ve published a new beta release, `v1.2.0-beta.1`, where I’ve decided to treat this as a breaking change. This is now the default (and only) method for sorting permissions.

Could you verify for me if it's fixed on your end? Would be great to hear whether it's resolved or if you're still running into problems.

If you're using the [ALCops VS Code extension](https://marketplace.visualstudio.com/items?itemName=Arthurvdv.alcops), you can switch to the beta channel in the extension settings. To force an update check, run the `ALCops: Check for Updates` command from the command palette.

**Comment by Arthurvdv** [comment]
The new sorting is now available with the the release of `v1.2.0` of ALCops.

Would be great to hear if it matches the AZ Dev Tools sorting. If everything is working as expected, feel free to close this issue.

**Comment by NKarolak** [comment]
@Arthurvdv At least for me: working as expected :-)

**Comment by guidorobben** [comment]
Works when there is no #pragma in the permission list

**Comment by Arthurvdv** [comment]
@guidorobben Yep, this is my design. Since AZ also does this explicitly, I’ve taken the exact same approach for ALCops here.
